core/extensions.rs

```diff
@@ -0,0 +1,136 @@
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use crate::error::AnyError;
+use crate::{OpFn, OpId, OpState};
+
+pub type SourcePair = (&'static str, &'static str);
+pub type OpPair = (&'static str, Box<OpFn>);
+pub type RcOpRegistrar = Rc<RefCell<dyn OpRegistrar>>;
+pub type OpMiddlewareFn = dyn Fn(&'static str, Box<OpFn>) -> Box<OpFn>;
+pub type OpStateFn = dyn Fn(&mut OpState) -> Result<(), AnyError>;
+
+#[derive(Default)]
+pub struct Extension {
+  js_files: Option<Vec<SourcePair>>,
+  ops: Option<Vec<OpPair>>,
+  opstate_fn: Option<Box<OpStateFn>>,
+  middleware_fn: Option<Box<OpMiddlewareFn>>,
+}
+
+impl Extension {
+  pub fn new(
+    js_files: Option<Vec<SourcePair>>,
+    ops: Option<Vec<OpPair>>,
```

Comment:
What is the distinction between `None` and `Some(vec![])`  ?

Reply:
Practically speaking there's no meaningful distinction, it's an implementation detail that will evolve and shouldn't be user facing but one upside of `Option` is that if the values need to be consumed there's a clean `.take()` operation